ms-oidc/src/main/java/net/trajano/ms/oidc/AuthenticationUriBuilder.java

```diff
@@ -0,0 +1,109 @@
+package net.trajano.ms.oidc;
+
+import java.net.URI;
+import java.text.ParseException;
+
+import javax.annotation.PostConstruct;
+import javax.ws.rs.core.UriBuilder;
+
+import org.jose4j.jwt.JwtClaims;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cache.Cache;
+import org.springframework.cache.CacheManager;
+import org.springframework.stereotype.Component;
+
+import net.trajano.ms.auth.token.GrantTypes;
+import net.trajano.ms.auth.token.OAuthTokenResponse;
+import net.trajano.ms.auth.util.HttpAuthorizationHeaders;
+import net.trajano.ms.authz.spi.ClientValidator;
+import net.trajano.ms.core.CryptoOps;
+import net.trajano.ms.core.ErrorCodes;
+import net.trajano.ms.oidc.internal.HazelcastConfiguration;
+import net.trajano.ms.oidc.internal.IssuerConfig;
+import net.trajano.ms.oidc.internal.ServerState;
+import net.trajano.ms.oidc.internal.ServiceConfiguration;
+
+/**
+ * Provides a method to build an authentication URI.
+ *
+ * @author Archimedes Trajano
+ */
+@Component
+public class AuthenticationUriBuilder {
+
+    @Autowired
+    private ClientValidator clientValidator;
+
+    @Autowired
+    private CacheManager cm;
+
+    @Autowired
+    private CryptoOps cryptoOps;
+
+    private Cache nonceCache;
+
+    @Value("${realmName:client_credentials}")
+    private String realmName;
+
+    @Autowired
+    private ServiceConfiguration serviceConfiguration;
+
+    /**
+     * This builds the authentication URI.
+     *
+     * @param state
+     *            client state
+     * @param issuerId
+     *            issuer ID
+     * @param authorization
+     *            Authorization header (for the client ID and client secret)
+     * @param additionalClaims
+     *            additional claims.
+     * @return URI to authenticate with the IP.
+     */
+    public URI build(final String state,
+        final String issuerId,
+        final String authorization,
+        final JwtClaims additionalClaims) {
+
+        if (issuerId == null) {
+            throw OAuthTokenResponse.badRequest(ErrorCodes.INVALID_REQUEST, "Missing issuer_id");
+        }
+        if (state == null) {
+            throw OAuthTokenResponse.badRequest(ErrorCodes.INVALID_REQUEST, "Missing state");
+        }
+
+        final IssuerConfig issuerConfig = serviceConfiguration.getIssuerConfig(issuerId);
+        if (issuerConfig == null) {
+            throw OAuthTokenResponse.badRequest(ErrorCodes.INVALID_REQUEST, "Invalid issuer_id");
+        }
+
+        // TODO validate whether the client credentials are valid.
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

Reply:
Part of #23